plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/validation/ModelValidator.java

```diff
@@ -23,14 +25,73 @@
  * Validates a list of {@link GModelElement}s based on a set of validation rules and returns corresponding issue
  * {@link Marker}s.
  * An issue marker is a serializable description of the validation violation that can be visualised by the GLSP client.
+ *
+ * There are two default reasons for a validation:
+ * <ol>
+ * <li><i>live</i> validation is executed on open and after each operation. These validation rules should be rather
+ * fast. Overwrite {@link #doLiveValidation(GModelElement)} to implement the live validation rules.</li>
+ * <li><i>batch</i> validation is executed on demand by the client. These validation rules can be more expensive.
+ * Overwrite {@link #doBatchValidation(GModelElement)} to implement the batch validation rules.</li>
+ * </ol>
  */
 public interface ModelValidator {
 
    /**
     * Validates the given list of {@link GModelElement}s and returns a list of {@link Marker}s.
     *
     * @param elements The list of {@link GModelElement} to validate.
+    * @param reason   The reason for a validation request, such as "batch" or "live" validation.
     * @return A list of {@link Marker}s for the validated {@link GModelElement}s.
     */
-   List<Marker> validate(GModelElement... elements);
+   default List<Marker> validate(final List<GModelElement> elements, final String reason) {
+      List<Marker> markers = new ArrayList<>();
+
+      for (GModelElement element : elements) {
+         if (MarkersReason.LIVE.equals(reason)) {
+            markers.addAll(doLiveValidation(element));
+         } else {
+            markers.addAll(doBatchValidation(element));
+         }
+         if (element.getChildren() != null) {
```

Comment:
I think this is a unnecessary null-check. List properties of EObjects are always defined and should never be null. We should rather check if the children list is not empty.